bigtable/google/cloud/bigtable/row_data.py

```diff
@@ -421,7 +421,7 @@ def _copy_from_previous(self, cell):
                 cell.row_key = previous.row_key
             if not cell.family_name:
                 cell.family_name = previous.family_name
-            if not cell.qualifier:
+            if cell.qualifier is None: # Note: qualifier can be empty string
```

Comment:
This line fixes the new test but causes failures in others such as `TestPartialRowsData_JSON_acceptance_tests.test_two_unsplit_cells` and `TestPartialRowsData_JSON_acceptance_tests.test_with_labels`.

Reply:
Yes the famous [`None`-check][1]!

When would `cell.qualifier` (it's not actually a `google.cloud.bigtable.row_data.Cell`, it's a `google.cloud.bigtable.row_data.PartialCellData` believe) be set to `None`?

[1]: http://iruntheinternet.com/lulzdump/images/nunchucks-nuns-chuck-norris-nun-12593488599.jpg?id=

Reply:
As the server sends chunks of a row, it only sends the row key/column family/column qualifier for a new cell if it is different than the last cell (as an optimization). So if you receive the same column at two different timestamps the second cell won't have a qualifier set and will fall back to the previously read one.

Reply:
@garye Thanks for the info, I meant where in this codebase though :grinning:

Reply:
haha, I'm afraid I don't have a great answer for that!

Reply:
So I'm pretty sure [this][1] is the only place:

```python
cell = self._cell = PartialCellData(
    chunk.row_key,
    chunk.family_name.value,
    chunk.qualifier.value,
    chunk.timestamp_micros,
    chunk.labels,
    chunk.value)
```

But since we're accessing `chunk.qualifier.value` in on the protobuf I'm fairly certain it will never be `None` (checking now).

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/99db799c5ad40c272fb921b9b9ed754958c33d6a/bigtable/google/cloud/bigtable/row_data.py#L286-L292

Reply:
OK I confirmed this value can never be `None` unless we change how a `PartialCellData` instance is constructed:

```python
>>> from google.cloud.bigtable._generated import bigtable_pb2
>>> chunk = bigtable_pb2.ReadRowsResponse.CellChunk()
>>> chunk.qualifier.value == b''
True
>>> chunk.HasField('qualifier')
False
>>> chunk.qualifier is None
False
```